src/expire.c

```diff
@@ -224,7 +225,12 @@ void activeExpireCycle(int type) {
     /* Try to smoke-out bugs (server.also_propagate should be empty here) */
     serverAssert(server.also_propagate.numops == 0);
 
-    for (j = 0; j < dbs_per_call && timelimit_exit == 0; j++) {
+    /* Stop iteration when one of the following conditions is met:
+     *
+     * 1) We have checked a sufficient number of databases with expiration time.
+     * 2) The time limit has been exceeded.
+     * 3) All databases have been traversed. */
+    for (j = 0; dbs_performed < dbs_per_call && timelimit_exit == 0 && j < server.dbnum; j++) {
```

Comment:
maybe `j` should be limited to dbs_per_call*10 instead of dbnum? or the lower of the two?

Reply:
If all databases have a few voltile keys but most of the are in db 127, we will never reach those even after this change, right?

Some other options:

* We can start j at a random index between 0 and dbnum and wrap around modulo dbnum. Then, at least there is a chance we scan all dbs some time. This will make active expire more fair between the dbs.

* We can count the size of each expire dict and limit that insted of limiting the number of dbs scanned.

Reply:
we already have a index: `current_db`.

Reply:
i don't know how many dbs, but if there are a lot, and most are with non-volatile keys, then this loop can run long time (if someone has more than 16db, maybe they also have more than 16384 :shrug:)
reminds me of what we have in SCAN, skipping empty slots up to a limit.

Reply:
>skipping empty slots up to a limit

Adding a new BIT for databases? Not sure whether it is worth it and the performance loss incurred.

Reply:
not sure what you mean.
i just meant to change the exit condition of that loop to depend on `MIN(server.dbnum, dbs_per_call * 10)`

Reply:
To summarize IIUC: For slots, we have a BIT (binary indexed tree) to find non-empty dicts without looping over them. That'd work for databases too but it's overkill. Then, if looping over all databases is too heavy, then `MIN(server.dbnum, dbs_per_call * 10)` is a simple and good enough compromise?

Reply:
maybe `MIN(server.dbnum, dbs_per_call * dbs_per_call)`?

Reply:
why?
considering a case where less than 10% of the dbs have volatile keys, do you wanna scan more?

Reply:
I did a simple test to measure the average time it takes to loop through the databases when there are no expires data:

databases | avg active expire time (us)
-- | --
1 | 0.16
16 | 0.43
64 | 1.25
160 | 2.87
256 | 3.77
1024 | 14.81
4096 | 57.80
16384 | 207.11

we can choose a reasonable threshold.

Reply:
ok, for someone with 16k databases, and only one with a single volatile key, this look will take just 0.2ms?
i guess that's fine.